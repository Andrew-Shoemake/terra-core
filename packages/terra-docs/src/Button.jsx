import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import ThemeContext from 'terra-theme-context';

import styles from './Button.module.scss';

const KEY_RETURN = 13;
const KEY_SPACE = 32;

const cx = classNames.bind(styles);

/**
 * Enables focus styles for the target of the given event. Typically used as an onBlur callback on selectable elements.
 */
const enableFocusStyles = (event) => {
  event.currentTarget.setAttribute('data-focus-styles-enabled', 'true');
};

/**
 * Disables focus styles for the target of the given event. Typically used as an onMouseDown callback on selectable elements.
 */
const disableFocusStyles = (event) => {
  event.currentTarget.setAttribute('data-focus-styles-enabled', 'false');
};

const propTypes = {
  /**
  * The main body text of the notice.
  */
  children: PropTypes.string,
  /**
   * Callback function triggered when key is pressed.
   */
  onKeyDown: PropTypes.func,
};

const Button = ({
  onKeyDown,
  children,
  ...customProps
}) => {
  const theme = React.useContext(ThemeContext);

  const buttonClassNames = classNames(
    cx([
      'button',
      theme.className,
    ]),
    customProps.className,
  );

  const handleKeyDown = (event) => {
    if (event.nativeEvent.keyCode === KEY_SPACE || event.nativeEvent.keyCode === KEY_RETURN) {
      if (onKeyDown) {
        onKeyDown(event);
      }
    }
  };

  return (
    <button
      {...customProps}
      type="button"
      className={buttonClassNames}
      onKeyDown={event => handleKeyDown(event)}
      onBlur={enableFocusStyles}
      onMouseDown={disableFocusStyles}
      data-focus-styles-enabled
    >
      {children}
    </button>
  );
};

Button.propTypes = propTypes;

export default Button;
