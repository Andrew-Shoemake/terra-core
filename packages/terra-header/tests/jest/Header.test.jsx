/* eslint-disable no-alert, no-console */
import React from 'react';
import Button from 'terra-button';
/* eslint-disable-next-line import/no-extraneous-dependencies */
import Header from '../../src/Header';

describe('Header', () => {
  const title = 'title';
  const level = 2;
  const content = <Button text="Button" />;
  // Snapshot Tests
  it('should render a default header', () => {
    const header = <Header />;
    const wrapper = shallow(header);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a header with given title', () => {
    const header = <Header title={title} />;
    const wrapper = shallow(header);
    expect(wrapper).toMatchSnapshot();
  });

  it('should fail if title is not a string', () => {
    expect(() => {
      shallow(<Header title={1} />);
    }).toThrowError();
    expect(() => {
      shallow(<Header title={<Button text={title} />} />);
    }).toThrowError();
  });

  it('should render a default header', () => {
    const header = <Header title={title} level={level} />;
    const wrapper = shallow(header);
    expect(wrapper).toMatchSnapshot();
  });

  it('should fail if level is not a number between 1 and 6', () => {
    expect(() => {
      shallow(<Header title={title} level={0} />);
    }).toThrowError();
    expect(() => {
      shallow(<Header title={title} level={7} />);
    }).toThrowError();
    expect(() => {
      shallow(<Header title={title} level="5" />);
    }).toThrowError();
    expect(() => {
      shallow(<Header title={title} level={<Button text="5" />} />);
    }).toThrowError();
  });

  it('should render a header with startContent', () => {
    const header = <Header startContent={content} />;
    const wrapper = shallow(header);
    expect(wrapper).toMatchSnapshot();
  });

  it('should fail if startContent is not an element', () => {
    expect(() => {
      shallow(<Header startContent={0} />);
    }).toThrowError();
    expect(() => {
      shallow(<Header startContent="content" />);
    }).toThrowError();
  });

  it('should render a header with endContent', () => {
    const header = <Header endContent={content} />;
    const wrapper = shallow(header);
    expect(wrapper).toMatchSnapshot();
  });

  it('should fail if endContent is not an element', () => {
    expect(() => {
      shallow(<Header endContent={0} />);
    }).toThrowError();
    expect(() => {
      shallow(<Header endContent="content" />);
    }).toThrowError();
  });

  it('should render a header with children', () => {
    const header = (
      <Header>
        {content}
      </Header>
    );
    const wrapper = shallow(header);
    expect(wrapper).toMatchSnapshot();
  });
});
