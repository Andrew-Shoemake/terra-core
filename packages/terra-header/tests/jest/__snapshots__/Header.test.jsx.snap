// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render a default header 1`] = `
<div
  className="flex-header"
>
  <div
    className="flex-fill"
  />
</div>
`;

exports[`Header should render a default header 2`] = `
<div
  className="flex-header"
>
  <div
    className="flex-fill"
  >
    <div
      className="title-container"
    >
      <h2
        className="title"
      >
        title
      </h2>
    </div>
  </div>
</div>
`;

exports[`Header should render a header with children 1`] = `
<div
  className="flex-header"
>
  <div
    className="flex-fill"
  />
  <Button
    className="flex-collapse"
    isBlock={false}
    isCompact={false}
    isDisabled={false}
    isIconOnly={false}
    isReversed={false}
    key=".0"
    text="Button"
    type="button"
    variant="neutral"
  />
</div>
`;

exports[`Header should render a header with endContent 1`] = `
<div
  className="flex-header"
>
  <div
    className="flex-fill"
  />
  <div
    className="flex-end end-group"
  >
    <Button
      isBlock={false}
      isCompact={false}
      isDisabled={false}
      isIconOnly={false}
      isReversed={false}
      text="Button"
      type="button"
      variant="neutral"
    />
  </div>
</div>
`;

exports[`Header should render a header with given title 1`] = `
<div
  className="flex-header"
>
  <div
    className="flex-fill"
  >
    <div
      className="title-container"
    >
      <h1
        className="title"
      >
        title
      </h1>
    </div>
  </div>
</div>
`;

exports[`Header should render a header with startContent 1`] = `
<div
  className="flex-header"
>
  <div
    className="flex-end start-group"
  >
    <Button
      isBlock={false}
      isCompact={false}
      isDisabled={false}
      isIconOnly={false}
      isReversed={false}
      text="Button"
      type="button"
      variant="neutral"
    />
  </div>
  <div
    className="flex-fill"
  />
</div>
`;
